sort dicts before calculating hashes

azul opened the pull request:
Ran into this issue using Claimchain with a custom FileStore
backend.

We store the content of blocks in files and then load them.
Loading them requires instantiating the block from the values.
Since dict order is non-deterministic and the hid seems to
depend on it this leads to errors.

Includes a test.

azul committed:
failing: test usage of hashes when copying Block

Ran into this issue using Claimchain with a custom FileStore
backend.

We store the content of blocks in files and then load them.
Loading them requires instantiating the block from the values.
Since dict order is non-deterministic and the hid seems to
depend on it this leads to errors.

This is just a failing test for now.

azul committed:
sort dicts before calculating hashes

Dicts have no defined order.
Therefore the hid of a Block should be independent of dict order.
That is two Blocks who are the same except for dict order
should have the same hid.

We achieve this by replacing all dicts with sorted lists
in the Block.items datastructure.
It's the only one that contains dicts.

azul commented:
@bogdan-kulynych could you have a look?

bogdan-kulynych commented on hippiehug-package/hippiehug/Chain.py:
```diff
@@ -14,6 +14,17 @@ def check_hash(key, val):
 
+def sort_dicts(unsorted):
+    if isinstance(unsorted, dict):
+        return sorted({k: sort_dicts(v) for k,v in unsorted.items()})
```
Particularly, this expression drops the values of the dict:
```python
In [1]: sorted({'a': 1, 'b': 2})
Out[1]: ['a', 'b']
```

azul replied:
Good catch!

bogdan-kulynych commented on hippiehug-package/tests/test_chain.py:
```diff
@@ -19,6 +20,28 @@ def test_block_duplicated():
+    }
+    b0 = Block(items=[cc], index=0, fingers=[], aux=42)
+    b1 = Block(b0.items, b0.index, b0.fingers, b0.aux)
+    assert b0.hid == b1.hid
```
You probably should test for the case `b0.hid != b1.hid` when some of the dict values differ --- to account for the behaviour of `sorted` mentioned above.

azul replied:
Thanks. Will start with that test and then make it pass.

bogdan-kulynych reviewed (commented):
`sort_dicts` drops the dict values:

```python
In [1]: sort_dicts({'a': 1, 'b': 2})
Out[1]: ['a', 'b']
```

It does not seem like intended behaviour. Is it?

azul committed:
make sure to include values in hash calculation

sorted({...}) only includes the keys.
sorted({...}.items()} works.

Includes a test.

azul commented:
@bogdan-kulynych I think i fixed the issue you described.

bogdan-kulynych reviewed (approved):
LGTM
